internal/trace/listener.go

```diff
@@ -11,19 +11,144 @@ package trace
 func (l *Listener) HandlerStarted(ctx context.Context, msg json.RawMessage) context.Context {
-	ctx, _ = ExtractTraceContext(ctx, msg)
+	ctx, _ = ContextWithTraceContext(ctx, msg)
+
+	if l.DDTraceEnabled {
+		tracer.Start(
+			tracer.WithService("aws.lambda"),
+			tracer.WithLambdaMode(true),
+			tracer.WithDebugMode(true),
+			tracer.WithGlobalTag("__dd.origin", "lambda"),
+		)
+
+		startFunctionExecutionSpan(
+			ctx,
+			l.MergeXrayTraces,
+		)
+
+		// Add the span to the context so the user can create child spans
+		ctx = tracer.ContextWithSpan(ctx, functionExecutionSpan)
+	}
+
 	return ctx
 }
 
-// HandlerFinished implemented as part of the HandlerListener interface
+// HandlerFinished finishes the function execution span (if it was started) and stops the tracer
 func (l *Listener) HandlerFinished(ctx context.Context) {
+	if functionExecutionSpan != nil {
+		functionExecutionSpan.Finish()
+	}
+
+	tracer.Stop()
+}
+
+func startFunctionExecutionSpan(ctx context.Context, mergeXrayTraces bool) {
+	// Extract information from context
+	lc, _ := lambdacontext.FromContext(ctx)
+	isColdStart := ctx.Value("cold_start").(bool)
+	var traceSource string
+	traceContext, ok := ctx.Value(traceContextKey).(map[string]string)
+	if ok {
+		traceSource = traceContext[sourceType]
+	} else {
+		logger.Error(fmt.Errorf("Error extracting Datadog trace context from context"))
+	}
+
+	functionArn := lc.InvokedFunctionArn
+	functionArn = strings.ToLower(functionArn)
+
+	// Separate version from rest of function ARN
+	parts := strings.Split(functionArn, ":")
+	functionVersion := "$LATEST"
+	if len(parts) > 7 {
+		functionArn = strings.Join(parts[0:7], ":")
+		functionVersion = parts[7]
+	}
+
+	// The function execution span must be made a child of the current span if the trace context came from an event OR merge X-Ray traces is enabled
+	// In other words, if merge X-Ray traces is NOT enabled and the trace context came from X-Ray, we should NOT make the execution span a child of the X-Ray span
+	var parentSpanContext ddtrace.SpanContext
+	if (traceSource == fromEvent) || mergeXrayTraces {
+		convertedSpanContext, err := convertTraceContextToSpanContext(traceContext)
+		if err == nil {
+			parentSpanContext = convertedSpanContext
+		}
+	}
+
+	span := tracer.StartSpan(
+		"aws.lambda",
+		tracer.SpanType("serverless"),
+		tracer.ChildOf(parentSpanContext),
+		tracer.ResourceName(lambdacontext.FunctionName),
+		tracer.Tag("cold_start", strconv.FormatBool(isColdStart)),
+		tracer.Tag("function_arn", functionArn),
+		tracer.Tag("function_version", functionVersion),
+		tracer.Tag("request_id", lc.AwsRequestID),
+		tracer.Tag("resource_names", lambdacontext.FunctionName),
+	)
+
+	if traceSource == fromXray && mergeXrayTraces {
+		span.SetTag("_dd.parent_source", traceSource)
+	}
+
+	functionExecutionSpan = span
+}
+
+// Convert trace context headers to a SpanContext object
+func convertTraceContextToSpanContext(traceContext map[string]string) (ddtrace.SpanContext, error) {
```

Comment:
It may be worth having a type for your `traceContext`, like `type traceContext map[string]string`.
That will make it more clear that functions like `unmarshalEventForTraceContext`, `addTraceContextToXRay`, `GetTraceHeaders`, etc. are all operating on the same data type, not just any old `map[string]string`. 
Plus, functions like these can become methods with cleaner names, like
```
func (c traceContext) toSpanContext() (ddtrace.SpanContext, error) {...}
```
What do you think?

Reply:
Great idea